getPullRequestEditedDiffRanges: use GitHub API

cklin opened the pull request:
This PR changes `getPullRequestEditedDiffRanges` so that it no longer obtains the PR diff through Git.  Instead, it now obtains the PR diff through the GitHub [Compare two commits](https://docs.github.com/en/rest/commits/commits?apiVersion=2022-11-28#compare-two-commits) API.

henrymercer reviewed (approved):
Thanks, a bunch of comments, but nothing major.

cklin commented on src/analyze.ts:
```diff
@@ -289,107 +302,162 @@ interface DiffThunkRange {
 ): Promise<DiffThunkRange[] | undefined> {
-  const checkoutPath = actionsUtil.getOptionalInput("checkout_path");
-  if (checkoutPath === undefined) {
+  const filediffs = await getFileDiffsWithBasehead(baseRef, headLabel, logger);
```
Done.

cklin commented on src/analyze.ts:
```diff
@@ -289,107 +302,162 @@ interface DiffThunkRange {
+interface FileDiff {
+  filename: string;
+  changes: number;
+  patch?: string | undefined;
```
Done.

cklin commented on src/analyze.ts:
```diff
@@ -289,107 +302,162 @@ interface DiffThunkRange {
+  if (fileDiff.patch === undefined) {
+    if (fileDiff.changes === 0) {
+      // There are situations where a changed file legitimately has no diff.
+      // For example, the file may be a binary file, or that the file may have
+      // been renamed with no changes to its contents. In these cases, the
```
The general principle is that diff-informed queries should preserve only those alerts that will be displayed on the PR, and the PR page displays an alert only if it can do so on the diff view.  Since the PR page does not display a diff for a renamed file, it will not show any alert in that file, so it is correct behavior not to preserve any alert in that renamed file.

Similarly for a binary file: the PR page will not display an alert in a binary file (even if there is one) because there is no way to place the alert on a diff line, so there is no need to preserve any alerts in a binary file.

cklin commented on src/analyze.ts:
```diff
@@ -289,107 +302,162 @@ interface DiffThunkRange {
+  // always terminate the last range of added lines.
+  lines.push(" ");
+
+  for (const line of lines) {
```
Done.

cklin commented on src/analyze.ts:
```diff
@@ -289,107 +302,162 @@ interface DiffThunkRange {
+  );
+
+  let currentLine = 0;
+  let additionRangeStartLine: number | undefined = undefined;
```
Done.

cklin commented:
@henrymercer I have addressed all comments; PTAL.

I want to thank you especially for suggesting that I add tests, because one of those tests caught a bug (incorrect path separator on Windows).  I have fixed the bug in the latest force push.

BLOCkCHAINMASTERINC commented:
Will I be able to fork my depreciated/deleted repository?...

On Mon, Jan 13, 2025, 2:47 PM Chuan-kai Lin ***@***.***>
wrote:

> @henrymercer <https://github.com/henrymercer> I have addressed all
> comments; PTAL.
>
> I want to thank you especially for suggesting that I add tests, because
> one of those tests caught a bug (incorrect path separator on Windows). I
> have fixed the bug in the latest force push.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/github/codeql-action/pull/2679#issuecomment-2588385598>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BM7HYNVVAKGPQVNO2POQMFL2KQ7A7AVCNFSM6AAAAABU5DUMQSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKOBYGM4DKNJZHA>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

henrymercer commented on src/analyze.ts:
```diff
@@ -764,3 +764,7 @@ export async function runCleanup(
   logger.endGroup();
 }
+
+export const exportedForTesting = {
+  getDiffRanges,
+};
```
I like this pattern, we should use it more in the rest of the Action :)
